lib/msf/core/post/windows/mssql.rb

```diff
@@ -109,13 +109,27 @@ def check_sqlcmd
         # @param [String] query the query to execute
         # @param [String] instance the SQL instance to target
         # @param [String] server the SQL server to target
+        # @param [String] database the database to connect to upon successfully authenticating
+        # @param [String] username the username to authenticate as
+        # @param [String] password the password to authenticate with
         # @return [String] the result of query
-        def run_sql(query, instance = nil, server = '.')
+        def run_sql(query, instance = nil, server = '.', database: nil, username: nil, password: nil)
           target = server
           if instance && instance.downcase != 'mssqlserver'
             target = "#{server}\\#{instance}"
           end
-          cmd = "#{@sql_client} -E -S #{target} -Q \"#{query}\" -h-1 -w 200"
+          cmd = "#{@sql_client}"
+          cmd += " -d #{database}" unless database.nil? || database.empty?
+          if username && password
```

Comment:
It's actually valid to only specify a password:

```
C:\Users\smcintyre>sqlcmd -S . -Q "select auth_scheme from sys.dm_exec_connections where session_id=@@spid" -h -1 -w 200 -P Password1!
SQL

(1 rows affected)

C:\Users\smcintyre>
```

In this case my user is `smcintyre` which maps to a `smcintyre` user using SQL authentication with the password of Password1!. Point is that in this case though, we don't need to specify a username, so this requirement isn't necessary.

It is however necessary to provide a password if a username is specified otherwise it'll prompt for the value which will likely result in the shell becoming defunct.

```
C:\Users\smcintyre>sqlcmd -S . -Q "select auth_scheme from sys.dm_exec_connections where session_id=@@spid" -h -1 -w 200 -U smcintyre
Password:
```

In this case, where a username is specified without a password, I'd recommend raising an exception. Without raising an exception, the calling module will assume the returned string is the results of the SQL query and won't know that the query was never executed because there was a problem. If the keyword argument is not `nil`, then it was most likely provided by the module author and thus this would be a problem the author of the calling code should address themselves.

Note that a blank password, does not cause the prompt to be displayed and is actually valid.

```
C:\Users\smcintyre>sqlcmd -S . -Q "select auth_scheme from sys.dm_exec_connections where session_id=@@spid" -h -1 -w 200 -U blank -P ""
SQL

(1 rows affected)

C:\Users\smcintyre>
```

Reply:
Thanks Spencer, appreciate you taking the time to test this out, next commit should contain changes based on these results 👍